opnode/node/api.go

```diff
@@ -0,0 +1,274 @@
+	HeadBlockHash Bytes32 `json:"headBlockHash"`
+	// safe block hash in the canonical chain
+	SafeBlockHash Bytes32 `json:"safeBlockHash"`
+	// block hash of the most recent finalized block
+	FinalizedBlockHash Bytes32 `json:"finalizedBlockHash"`
+}
+
+type ForkchoiceUpdatedStatus string
+
+const (
+	// given payload is valid
+	UpdateSuccess ForkchoiceUpdatedStatus = "SUCCESS"
+	// sync process is in progress
+	UpdateSyncing ForkchoiceUpdatedStatus = "SYNCING"
+)
+
+type ForkchoiceUpdatedResult struct {
+	// the result of the payload execution
+	Status ForkchoiceUpdatedStatus `json:"status"`
+	// the payload id if requested
+	PayloadID PayloadID `json:"payloadId"`
+}
+
+func GetPayload(ctx context.Context, cl *rpc.Client, log Logger, payloadId PayloadID) (*ExecutionPayload, error) {
+	e := log.New("payload_id", payloadId)
+	e.Debug("getting payload")
+	var result ExecutionPayload
+	err := cl.CallContext(ctx, &result, "engine_getPayloadV1", payloadId)
+	if err != nil {
+		e = log.New("payload_id", "err", err)
+		if rpcErr, ok := err.(rpc.Error); ok {
+			code := ErrorCode(rpcErr.ErrorCode())
+			if code != UnavailablePayload {
+				e.Warn("unexpected error code in get-payload response", "code", code)
+			} else {
+				e.Warn("unavailable payload in get-payload request")
+			}
+		} else {
+			e.Error("failed to get payload")
+		}
+		return nil, err
+	}
+	e.Debug("Received payload")
+	return &result, nil
+}
+
+func ExecutePayload(ctx context.Context, cl *rpc.Client, log Logger, payload *ExecutionPayload) (*ExecutePayloadResult, error) {
+	e := log.New("block_hash", payload.BlockHash)
+	e.Debug("sending payload for execution")
+	var result ExecutePayloadResult
+	err := cl.CallContext(ctx, &result, "engine_executePayloadV1", payload)
+	if err != nil {
+		e.Error("Payload execution failed", "err", err)
+		return nil, err
+	}
+	e.Debug("Received payload execution result", "status", result.Status, "latestValidHash", result.LatestValidHash, "message", result.ValidationError)
+	return &result, nil
+}
+
+func ForkchoiceUpdated(ctx context.Context, cl *rpc.Client, log Logger, state *ForkchoiceState, attr *PayloadAttributes) (ForkchoiceUpdatedResult, error) {
+	e := log.New("state", state, "attr", attr)
+	e.Debug("Sharing forkchoice-updated signal")
+
+	var result ForkchoiceUpdatedResult
+	err := cl.CallContext(ctx, &result, "engine_forkchoiceUpdatedV1", state, attr)
+	if err == nil {
+		e.Debug("Shared forkchoice-updated signal")
+		if attr != nil {
+			e.Debug("Received payload id", "payloadId", result.PayloadID)
+		}
+		return result, nil
+	} else {
+		e = e.New("err", err)
+		if rpcErr, ok := err.(rpc.Error); ok {
+			code := ErrorCode(rpcErr.ErrorCode())
+			e.Warn("Unexpected error code in forkchoice-updated response", "code", code)
+		} else {
+			e.Error("Failed to share forkchoice-updated signal")
+		}
+		return result, err
+	}
+}
+
+func BlockToPayload(bl *types.Block, random Bytes32) (*ExecutionPayload, error) {
```

Comment:
Do we expect to ever go from block to payload?

(That being said, not opposed to have this here in any case, it's pretty instructive and seems like a useful utility to have.)

Reply:
If we distribute block data in RLP encoded form, and want to insert it through the API, then it's useful to have. We can remove it later if we do not end up doing that. It's there for completeness of the API bindings.

Reply:
I think it would be useful to have a comment that says exactly this in the code!